### PR TITLE
Faster vars used tracking in simplify let visitor

### PR DESCRIPTION
**Pull request by abadams**
This visitor shows up as the main cost of lowering in very large
pipelines.

This visitor is for tracking which lets are actually used for real
inside the body of a let block (as opposed to the tracking we do when
mutating, which is approximate, because we could construct an Expr that
uses a Var and then discard it in a later mutation).

The old implementation made a map of all variables referenced, and then
checked each let name against that map one by one. If there are a small
number of lets outside a huge Stmt, this is bad, because the data
structure has to hold a number of names proportional to the Stmt size
instead of proportional to the number of lets.

This new implementation instead makes a hash set of the let names, and
than traverses the Stmt, removing names from the set as they are
encountered. This is a big speed-up.

We then make the speed-up larger by about the same factor again doing
the following:

1) Only add names to the map that might be used based on the recursive
mutate call. These are very very likely to be used, because we saw them
at least once, and mutations that remove *all* uses of a Var are rare.

2) The visitor should early out when the map becomes empty. The let
variables are often all used immediately, so this is frequent.

Speeds up lowering of local laplacian by 1.44x, 2.6x, and 4.8x
for 20, 50, and 100 pyramid levels respectively.

Speeds up lowering of resnet50 by 1.04x. 

Speeds up lowering of lens blur by 1.06x

Built on #8202 

Complementary with #8204 though the total speed-up from merging both is *not* going to be the product of the two speed-ups. The two PRs solve the same local laplacian pathological lowering time problem in two different ways.

### CODE REVIEW

**Comment by rootjalex** [comment]
It looks like this PR also contains #8202 ? Might need to merge with main to clean it up

File: src/Simplify_Let.cpp [hunk 1]
```diff
@@ -9,34 +11,50 @@ using std::vector;
...
+            unused_vars.erase(op->name);
+            IRVisitor::visit(op);
+        }
+    }
+
+    void visit(const Block *op) override {
+        // Early out at Block nodes if we've already seen every name we're
+        // interested in. In principal we could early-out at every node, but
+        // blocks, loads, and stores seem to be enough.
+        if (!unused_vars.empty()) {
```
Comment (rootjalex) [inline src/Simplify_Let.cpp:39]: Is it guaranteed that a block will never contain a let? If so, it seems like this method would not recursively simplify a block with lets

Comment (abadams) [reply src/Simplify_Let.cpp:39]: How so? This is just a visitor to check whether a stmt references a set of vars. This early out is when we're sure it references all of them. It's not a mutator. The body has already been simplified before this visitor is even run.

**Comment by abadams** [comment]
With the last merge, when combined with #8204, this is 6.7x faster for local laplacian with 100 pyramid levels. But Amdahl's law has kicked in here. For just the pass that was slow (the simplification pass just after vectorization, because it did the first simplification after unify_duplicate_lets), this is actually over 400x faster.

File: src/Simplify_Let.cpp [hunk 2]
```diff
@@ -226,14 +247,35 @@ Body Simplify::simplify_let(const LetOrLetStmt *op, ExprInfo *bounds) {
...
     // of writing, both cover cases that the other does not:
     // - var_info prevents duplicate lets from being generated, even
     //   from different Frame objects.
-    // - vars_used avoids dead lets being generated in cases where vars are
+    // - unused_vars avoids dead lets being generated in cases where vars are
     //   seen as used by var_info, and then later removed.
-    std::map<std::string, int> vars_used;
-    count_var_uses(result, vars_used);
+
+    std::unordered_set<std::string> unused_vars(frames.size() * 2);
```
Comment (rootjalex) [inline src/Simplify_Let.cpp:257]: Where does `frames.size() * 2` come from?

Comment (rootjalex) [reply src/Simplify_Let.cpp:257]: Oh, each frame can maybe insert two names. Maybe add a comment above this definition, for explanation?

Comment (abadams) [reply src/Simplify_Let.cpp:257]: Actually I think old_uses and new_uses might be mutually exclusive. Will test.

Comment (abadams) [reply src/Simplify_Let.cpp]: Yep. Fixed.

**Comment by rootjalex** [comment]
Is it reasonable to write tests for this new functionality, or do existing tests already stress it enough?

**Comment by abadams** [comment]
Everything breaks if you remove a let that you should not have. I also confirmed the existing tests fail if you leave too many lets in (it messes up some instruction selection tests).

**Comment by abadams** [comment]
Just needs an approval

**Review by rootjalex — APPROVED** [review]
Lgtm

**Pull request merged**